keps/sig-api-machinery/1630-kubeconfig-qps-burst/README.md

```diff
@@ -0,0 +1,130 @@
+We will expose the values already present in the [REST config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L114-L120), via a kubeconfig file inside the client stanza.
+We will wire the in-cluster kubeconfig file to honor the presence of a new file to allow customization inside a cluster at the discretion of a different, user provided, controller.
+
+## Motivation
+
+We've always had reasons to override these values, so we created one-off arguments (`--kube-api-burst`, `--kube-api-qps`).
+With widespread adoption of operators and other in-cluster workloads, having a standard way to provide these values makes sense.
+In addition, with the introduction of API priority and fairness, being able to configure and even disable this client-side rate limiting in compatible environments is important.
+
+### Goals
+
+1. Allow easy and consistent configuration for the existing client-side rate limiting.
+
+### Non-Goals
+
+1. Allow customization of client-side rate limiting strategies.
+
+## Proposal
+
+Add to the `AuthInfo` section of a kubeconfig.
+This section identifies a particular client identity.
+This identity spans all the contexts with different namespaces, so it is a closer logical match.
+You don't generally want different QPS characteristics based on namespace.
+
+```go
+type AuthInfo struct {
+	// ...
+	// QPS indicates the maximum QPS to the master from this client.
+	// If it's zero, the created RESTClient will use DefaultQPS: 5
+	QPS float32
+
+	// Maximum burst for throttle.
+	// If it's zero, the created RESTClient will use DefaultBurst: 10.
+	Burst int32
+}
+```
+
+### in-cluster kubeconfig overrides
+The in-cluster config is the automatic configuration that exists within pods.
+The client code itself can be written to consume an overrides.kubeconfig file, which will be merged with the standard
+kubeconfig merging rules that we use the env var path today.
+The controller code will *not* be updated.
+If a cluster-admin wants to control this file, he can provide his own controller to write content into this file in 
+serviceaccount token secrets.
+
+### Test Plan
+
+This is heavily unit testable.
+
+### Graduation Criteria
+
+The API itself is well-known since near the beginning of kube.
+We are simply exposing existing values, it will enter as GA.
+
+### Upgrade / Downgrade Strategy
+
+Old clients will ignore the unknown field.
+Old `kubectl` will rewrite the stanza and remove the new values.
+
+### Version Skew Strategy
+
+Old clients will ignore the unknown field.
+Old `kubectl` will rewrite the stanza and remove the new values.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+1. Connecting to the kube-apiserver to retrieve a value for default QPS and burst.
+   No default value is going to be good across all workloads.
+   It's also mechanicaly weird to create a client to get a client.
```

Comment:
I agree this is a bit weird, but it wouldn't have to be a separate request-- we could make apiserver set a header with its desired rate limit, and the client just obeys it after the first request. This might be simpler?

Reply:
I am actually liking this (return a header) more the longer I think about it. Since we have to recompile all clients anyway...

Reply:
Yeah.

Reply:
After more thought I think this is too heavyweight (on API changes -- but so is the main proposal), and it's easy to believe that not all clients would want the same rate limit.

Reply:
(So I guess this would still be in consideration if it were just about turning off the rate limit, not about setting it.)